### PR TITLE
UI changes to display custom metrics

### PR DESCRIPTION
**Pull request by anushree-n**
The function startCustomMetrics is called once to setup the custom graphs depending on the number of custom metrics.
The function drawCustomMetrics is called periodically to draw all the graphs of custom metrics.

### CODE REVIEW

File: pages/containers.go [hunk 1]
```diff
@@ -231,20 +231,21 @@ func serveContainersPage(m manager.Manager, w http.ResponseWriter, u *url.URL) e
...
-		Root:               rootDir,
+		DisplayName:            displayName,
+		ContainerName:          escapeContainerName(cont.Name),
+		ParentContainers:       parentContainers,
+		Subcontainers:          subcontainerLinks,
+		Spec:                   cont.Spec,
+		Stats:                  cont.Stats,
+		MachineInfo:            machineInfo,
+		IsRoot:                 cont.Name == "/",
+		ResourcesAvailable:     cont.Spec.HasCpu || cont.Spec.HasMemory || cont.Spec.HasNetwork || cont.Spec.HasFilesystem || cont.Spec.HasCustomMetrics,
```
Comment (anushree-n) [inline pages/containers.go]: Addition of customMetrics to ResourcesAvailable is probably not necessary, as cpu, memory, network, fs are the standard metrics which need to be present.

File: pages/static/containers_js.go [hunk 2]
```diff
@@ -468,6 +468,30 @@ function drawFileSystemUsage(machineInfo, stats) {
 	}
 }
 
+// Draw the graph for Custom Metrics.
```
Comment (anushree-n) [inline pages/static/containers_js.go]: Assumption: PollingFrequency is same for all custom metrics which leads to custom_metrics.length being same in all stats corresponding to different timestamps.

File: pages/containers.go (hunk 1 above)
Comment (rjnagal) [inline pages/containers.go]: ResourcesAvailable is used to draw the spec table. We can skip CustomMetrics as the details won't show up there.

File: pages/containers_html.go [hunk 3]
```diff
@@ -221,6 +221,16 @@ const containersHtmlTemplate = `
           </div>
         </div>
 	{{end}}
+        {{if .CustomMetricsAvailable}}
```
Comment (rjnagal) [inline pages/containers_html.go]: We should make this its own h3 heading as Application Metrics.

File: pages/static/containers_js.go [hunk 4]
```diff
@@ -468,6 +468,30 @@ function drawFileSystemUsage(machineInfo, stats) {
...
+			var elements = [];
+                	elements.push(stats.stats[i].timestamp);
+			if (cur.int_points.length != 0) {
+				elements.push(cur.int_points[0])
+			} else if (cur.float_points.length !=0) {
+				elements.push(cur.float_points[0])
+			}
+                	data.push(elements);
+        	}
+        	drawLineChart(titles, data, elementId+"-"+cur.name, "Value")
```
Comment (rjnagal) [inline pages/static/containers_js.go]: Why not just show cur.name ? Do we also get units in the config?

**Comment by rjnagal** [comment]
Let's hold on the UI changes a bit. I want to change our storage structure first.

The way we are storing stats is a bit inefficient.
The current structure stores name, type, label and int/float array for each stat point  (because we started to append metrics directly to containerStats).

I think we need to do two things:
1. take out name, type, label and push them only once in spec.
2. only store one struct {intval, floatval} per metric in container stats. We can add a map from metric name to this new struct. If we dont' want to waste storing two values for each metric, we can make two custom metrics map - one for int and one for float.

The other option is to not append custom metrics under ContainerStats, but that will need another timeStore.

I am going to start with separating out the spec first. let me know what you think @anushree-n @vmarmol

**Comment by anushree-n** [comment]
Append metrics under ContainerStats
     - Already in place
     - Logically , custom metrics follows cpu, memory stats
     - If polling frequency of custom metrics is higher than that of regular stats, containerStats can be uneven.

Different timeStore
     -Extra overhead
     - Handling of data(to display in API and UI) may be easier

**Comment by vmarmol** [comment]
@anushree-n is right about the different periods. Long term we want to move everything to metrics which would solve the timing problem and be the most efficient storage mechanism. That will probably take some time though :) @rjnagal can't we coalesce the name,labels into the same metric? The way it is exposed today it should handle multiple points with the same metadata. Maybe we need an intermediate structure as an output to the Collects.

File: pages/static/containers_js.go [hunk 5]
```diff
@@ -696,12 +705,67 @@ function refreshStats() {
...
+	if(metricsInfo.length == 0) {
+		return;
+	}
+
+	for (var containerName in metricsInfo) {
+		var container = metricsInfo[containerName];
+		var metricIndex = -1 ;
+		for (var metricName in container) {
+			metricIndex++;
+			var titles = ["Time", containerInfo.spec.custom_metrics[metricIndex].units];
```
Comment (rjnagal) [inline pages/static/containers_js.go]: The order of stats and spec doesn't match. In my tests, the units gets mixed up.
Maps in go don't have a fixed order. I'd suggest changing the loop to iterate over the spec and locate spec.custom_metrics[index].name in metrics map returned by appmetrics.

**Comment by anushree-n** [comment]
@rjnagal Let me know if any changes need to be done. Thanks!

**Comment by rjnagal** [comment]
This looks great.

LGTM.

Squash and lets commit.

**Comment by anushree-n** [comment]
Squashed. :)

**Pull request merged**